File: frontend/lib/models/audit_log.dart
class AuditLog {
  final int id;
  final String tableName;
  final int recordId;
  final String action;
  final Map<String, dynamic>? oldValues;
  final Map<String, dynamic>? newValues;
  final int? userId;
  final String? userName;
  final String? createdAt;

  AuditLog({
    required this.id,
    required this.tableName,
    required this.recordId,
    required this.action,
    this.oldValues,
    this.newValues,
    this.userId,
    this.userName,
    this.createdAt,
  });

  factory AuditLog.fromJson(Map<String, dynamic> json) {
    Map<String, dynamic>? parseValues(dynamic v) {
      if (v == null) return null;
      if (v is Map<String, dynamic>) return v;
      if (v is String) {
        try {
          return Map<String, dynamic>.from(Uri.parse(v) as Map? ?? {});
        } catch (_) {}
      }
      return null;
    }

    return AuditLog(
      id: json['id'] as int,
      tableName: json['table_name'] as String,
      recordId: json['record_id'] as int,
      action: json['action'] as String,
      oldValues: parseValues(json['old_values']),
      newValues: parseValues(json['new_values']),
      userId: json['user_id'] as int?,
      userName: json['user']?['name'] as String?,
      createdAt: json['created_at'] as String?,
    );
  }
}

File: ios/Runner/Models/AuditLog.swift
import Foundation

struct AuditLog: Decodable, Identifiable {
    let id: Int
    let tableName: String
    let recordId: Int
    let action: String
    let oldValues: [String: AnyCodable]?
    let newValues: [String: AnyCodable]?
    let userId: Int?
    let userName: String?
    let createdAt: String?

    private enum CodingKeys: String, CodingKey {
        case id
        case tableName  = "table_name"
        case recordId   = "record_id"
        case action
        case oldValues  = "old_values"
        case newValues  = "new_values"
        case userId     = "user_id"
        case user
        case createdAt  = "created_at"
    }

    private struct UserNested: Decodable { let name: String }

    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        id         = try c.decode(Int.self, forKey: .id)
        tableName  = try c.decode(String.self, forKey: .tableName)
        recordId   = try c.decode(Int.self, forKey: .recordId)
        action     = try c.decode(String.self, forKey: .action)
        oldValues  = try? c.decode([String: AnyCodable].self, forKey: .oldValues)
        newValues  = try? c.decode([String: AnyCodable].self, forKey: .newValues)
        userId     = try? c.decode(Int.self, forKey: .userId)
        createdAt  = try? c.decode(String.self, forKey: .createdAt)
        userName   = (try? c.decode(UserNested.self, forKey: .user))?.name
    }
}
